wiki/Writing-tests-with-good-behavioural-coverage.md

```diff
@@ -0,0 +1,1265 @@
+
+```
+@Test
+fun testWithdraw_withReceipt_generatesReceipt() {
+    val output = ByteArrayOutputStream()
+    System.setOut(PrintStream(output))
+
+    account.withdraw("user", "password", 200.0, needReceipt = true)
+
+    assertThat(output.toString().trim()).contains("Receipt: Withdrew 200.0. Current balance: 800.0")
+    System.setOut(System.out)
+}
+```
+
+**c. To verify with passbook details:**
+
+Confirms balance statement with the passbook file being updated when a file is provided.
+
+```
+@Test
+fun testWithdraw_withPassbook_updatesPassbook() {
+    val passbookFile = File("passbook.pdf")
+    passbookFile.createNewFile()
+    val output = ByteArrayOutputStream()
+    System.setOut(PrintStream(output))
+
+    account.withdraw("user", "password", 200.0, file = passbookFile)
+
+    // Read the passbook file and check its contents
+    val fileContents = passbookFile.readText()
+    assertThat(fileContents).contains("Withdrew 200.0. New balance is 800.0")
+
+    System.setOut(System.out)
+}
+```
+
+**d. To verify based on the log message:**
+
+Validates that the correct message about the withdrawal is logged.
+
+```
+@Test
+fun testWithdraw_withSufficientBalance_logsCorrectMessage() {
+    val output = ByteArrayOutputStream()
+    System.setOut(PrintStream(output))
+
+    account.withdraw("user", "password", 200.0)
+
+    assertThat(output.toString().trim()).contains("New balance is 800.0")
+    System.setOut(System.out)
+}
+```
+
+**e. To verify with the balance is updated correctly:**
+
+Ensures that the balance is updated correctly after a withdrawal.
+
+```
+@Test
+fun testWithdraw_withSufficientBalance_updatesBalanceCorrectly() {
+    account.withdraw("user", "password", 200.0)
+    assertThat(account.balance).isEqualTo(800.0)
+}
+```
+
+These tests cover various aspects of the withdrawal functionality, ensuring that the balance updates correctly with receipts, passbooks and output messages. Although the core functionality of withdrawing funds and updating the balance is consistent, it can be observed in multiple ways. Each test focuses on a specific verification method while ultimately validating the same core functionality.
+
+## Testing Practices for Effective Validation
+
+The following best practices are meant to assist in the organization and execution of tests. By following these guidelines, you can ensure that your testing process is efficient, thorough, and well-structured.
+
+### 1. Use of Helper Functions
+
+Helper functions are crucial for avoiding repetitive code, especially for pre-setup tasks that are used across multiple test cases. In this scenario, initializing the bank data or ensuring the proper state of the BankAccount object can be encapsulated in helper functions.
+
+```
+// Helper function to initialize a BankAccount with default data
+fun createDefaultBankAccount(): BankAccount {
+    return BankAccount(1000.0, "user", "password")
```

Comment:
Here & elsewhere: please use 2 space tabbing for consistency with the Oppia Android codebase.

Reply:
Done, updated every snippet to use 2-space tabbing.